src/utilities/geometry/Polygon3d.cpp

```diff
@@ -0,0 +1,236 @@
+*  disclaimer.
+*
+*  (2) Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following
+*  disclaimer in the documentation and/or other materials provided with the distribution.
+*
+*  (3) Neither the name of the copyright holder nor the names of any contributors may be used to endorse or promote products
+*  derived from this software without specific prior written permission from the respective party.
+*
+*  (4) Other than as required in clauses (1) and (2), distributions in any form of modifications or other derivative works
+*  may not use the "OpenStudio" trademark, "OS", "os", or any other confusingly similar designation without specific prior
+*  written permission from Alliance for Sustainable Energy, LLC.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER(S) AND ANY CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+*  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+*  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER(S), ANY CONTRIBUTORS, THE UNITED STATES GOVERNMENT, OR THE UNITED
+*  STATES DEPARTMENT OF ENERGY, NOR ANY OF THEIR EMPLOYEES, BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+*  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+*  STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+*  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***********************************************************************************************************************/
+
+#include "Polygon3d.hpp"
+#include "Vector3d.hpp"
+#include "Geometry.hpp"
+#include "Intersection.hpp"
+#include <utilities/geometry/Transformation.hpp>
+
+namespace openstudio {
+
+Polygon3d::Polygon3d() {}
+
+Polygon3d::Polygon3d(const Point3dVector& outerPath) {
+  for (const auto& p : outerPath) {
+    m_outerPath.emplace_back(p);
+  }
+}
+
+Polygon3d::Polygon3d(const Point3dVector& outerPath, const Point3dVectorVector& innerPaths) {
+  for (const auto& p : outerPath) {
+    m_outerPath.emplace_back(p);
+  }
+
+  for (const auto& innerPath : innerPaths) {
+    addHole(innerPath);
+  }
+}
+
+/// Adds a point to the polygon perimeter
+void Polygon3d::addPoint(const Point3d& point) {
+  m_outerPath.emplace_back(point);
+}
+
+/// Sets the perimeter of the polygon
+void Polygon3d::setOuterPath(const Point3dVector& outerPath) {
+  m_outerPath = outerPath;
+}
+
+Point3dVector Polygon3d::getOuterPath() const {
+  return m_outerPath;
+}
+
+Point3dVectorVector Polygon3d::getInnerPaths() const {
+  return m_innerPaths;
+}
+
+void Polygon3d::addHole(const Point3dVector& hole) {
+  m_innerPaths.emplace_back(hole);
```

Comment:
Should you check that hole is on the same plane as the outer path? We have a Plane class you could use

Reply:
Should you check the hole is contained in the outer path?